File: 카카오/2020 카카오 개발자 인턴십/키패드 누르기.cpp
#include <string>
#include <vector>

using namespace std;

string solution(vector<int> numbers, string hand) {
    string answer = "";
    pair<int,int> left = make_pair(4,1);
    pair<int,int> right = make_pair(4,3);
    pair<int,int> middle;
    int left_dis = 0;
    int right_dis = 0;
    
    for (int i = 0; i < numbers.size(); i++){
        left_dis = 0;
        right_dis = 0;
        
        if(numbers[i] == 1 || numbers[i] == 4 || numbers[i] == 7){
            left.first = numbers[i] / 3 + 1;
            left.second = 1;
            answer += "L";
        }
        else if(numbers[i] == 3 || numbers[i] == 6 || numbers[i] == 9){
            right.first = numbers[i] / 3;
            right.second = 3;
            answer += "R";
        }
        
        else{
            if(numbers[i] == 0){
                middle.first = 4;
                middle.second = 2;
            }
            else{
                middle.first = numbers[i]/3 + 1;
                middle.second = 2;
            }
            if(middle.first >= left.first){
                left_dis += middle.first - left.first;
            }
            else{
                left_dis += left.first - middle.first;
            }
            left_dis += middle.second - left.second;
            
            if(middle.first >= right.first){
                right_dis += middle.first - right.first;
            }
            else{
                right_dis += right.first - middle.first;
            }
            right_dis += right.second - middle.second;
            
            if(left_dis < right_dis){
                left.first = middle.first;
                left.second = middle.second;
                answer += "L";
            }
            else if(left_dis > right_dis){
                right.first = middle.first;
                right.second = middle.second;
                answer += "R";
            }
            else{
                if(hand == "right"){
                    right.first = middle.first;
                    right.second = middle.second;
                    answer += "R";
                }
                else{
                    left.first = middle.first;
                    left.second = middle.second;
                    answer += "L";
                    
                }
            }
            
        }
        
        
    }
    
    
    return answer;
}

File: 카카오/2020 카카오 개발자 인턴십/키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    
    #왼손 오른손 현재 위치
    left_position = [-1,3]
    right_position = [1,3]
    
    
    
    for i in numbers:
        
        
        if i == 1 or i == 4 or i== 7:
            answer += "L"
            left_position = [-1,int((i - 0.1)/3)]
        elif i == 3 or i == 6 or i == 9:
            answer += "R"
            right_position = [1,int((i - 0.1)/3)]
        else:
            if i == 0:
                middle_position = [0,3]
            else:
                middle_position = [0,int(i/3)]
            
            if abs(left_position[1] - middle_position[1]) + abs(left_position[0] - middle_position[0]) < abs(right_position[1] - middle_position[1]) + abs(right_position[0] - middle_position[0]):
                left_position = middle_position
                answer += "L"
            elif abs(left_position[1] - middle_position[1]) + abs(left_position[0] - middle_position[0]) > abs(right_position[1] - middle_position[1]) + abs(right_position[0] - middle_position[0]):
                answer += "R"
                right_position = middle_position
                
            else:
                if hand == "right":
                    answer += "R"
                    right_position = middle_position
                else:
                    answer += "L"
                    left_position = middle_position
    
    
    return answer
